cpp/src/neighbors/sample_filter.cuh

```diff
@@ -140,4 +146,89 @@ void bitmap_filter<bitmap_t, index_t>::to_csr(raft::resources const& handle, csr
   raft::sparse::convert::bitmap_to_csr(handle, bitmap_view_, csr);
 }
 
+struct ivf_to_sample_filter_dev {
+  FilterType tag;
+  const void* obj_ptr;
+
+  __device__ bool operator()(uint32_t q, uint32_t c, uint32_t s) const
+  {
+    switch (tag) {
+      case FilterType::None:
+        return reinterpret_cast<
+                 const filtering::ivf_to_sample_filter<int64_t, filtering::none_sample_filter>*>(
+                 obj_ptr)
+          ->operator()(q, c, s);
+      case FilterType::Bitset:
+        return reinterpret_cast<const filtering::ivf_to_sample_filter<
+          int64_t,
+          filtering::bitset_filter<uint32_t, int64_t>>*>(obj_ptr)
+          ->operator()(q, c, s);
+      default: return true;
+    }
+  }
+};
+
+template <typename dev_filter_t>
+__global__ void destruct_dev_filter(dev_filter_t* p_filter)
+{
+  p_filter->~dev_filter_t();
+}
+
+template <typename InnerFilterT, typename... Args>
+__global__ void init_inner_filter(InnerFilterT* p_inner_filter, Args... args)
+{
+  new (p_inner_filter) InnerFilterT(args...);
+}
+
+template <typename OuterFilterT, typename InnerFilterT, typename... Args>
+__global__ void init_outer_filter(OuterFilterT* p_outer_filter,
+                                  const int64_t* const* inds_ptrs,
+                                  InnerFilterT* p_inner_filter)
+{
+  new (p_outer_filter) OuterFilterT(inds_ptrs, *p_inner_filter);
+}
+
+/* Device side filter wrapper for ivf_to_sample_filter. ivf_to_sample_filter_dev is used to avoid
+ * dynamic dispatching on device. */
+template <typename InnerFilterT, typename... InnerArgs>
+struct ivf_to_sample_dev_wrapper {
+  using OuterFilterT = filtering::ivf_to_sample_filter<int64_t, InnerFilterT>;
+  OuterFilterT* p_outer_filter;
+  InnerFilterT* p_inner_filter;
+  filtering::ivf_to_sample_filter_dev h_placeholder;
+
+  ivf_to_sample_dev_wrapper(FilterType tag,
+                            const int64_t* const* inds_ptrs,
+                            InnerArgs... inner_args)
+  {
+    cudaMalloc(&p_inner_filter, sizeof(InnerFilterT));
+    init_inner_filter<<<1, 1>>>(p_inner_filter, inner_args...);
+    cudaDeviceSynchronize();
+
+    cudaMalloc(&p_outer_filter, sizeof(OuterFilterT));
+    init_outer_filter<<<1, 1>>>(p_outer_filter, inds_ptrs, p_inner_filter);
+    cudaDeviceSynchronize();
+
+    h_placeholder = {tag, p_outer_filter};
+  }
+
+  ~ivf_to_sample_dev_wrapper()
+  {
+    destruct_dev_filter<<<1, 1>>>(p_outer_filter);
```

Comment:
Call twice the kernel (or combine it) then cudaFree both. the sync is unnecessary I think

Reply:
Yes that make sense. The sync must be done on the stream not on the device so `stream_.synchronize()` would be better because it is calling  `cudaStreamSynchronize()`.